File: nishuProbs/easy/addBinary.js
/**
 * https://leetcode.com/problems/add-binary/
Given two binary strings a and b, return their sum as a binary string.

Example 1:
Input: a = "11", b = "1"
Output: "100"

Example 2:
Input: a = "1010", b = "1011"
Output: "10101"


Constraints:

1 <= a.length, b.length <= 10^4
a and b consist only of '0' or '1' characters.
Each string does not contain leading zeros except for the zero itself.

0 = 0
1 = 1
0 + 1 = 10
11 + 1 = 100

  1010
+ 1011
=10101

  1111
+ 1110
=11101

[1,0,1,0]
[1,0,1,1]
=
[10, 0, 10, 1]
IF current is zero and next number is 10,
remove the zero from result array

 * @param {string} a
 * @param {string} b
 * @return {string}
 */
/*
4 cases when adding
a = 1, b = 1 where result is 0 and carry the one
a = 0 b = 1 & a = 1 b = 0, result is 1
a, b = 0, result is zero
loop through a b backwards
IF a or b is undefined, take the other number and add to result string
*/
var _addBinary = function(a, b) {
    let result = '';
    let len = a.length > b.length ? a.length - 1 : b.length - 1;
    let sums = [];
    let as = a.split('');
    let bs = b.split('');
    for (let i = len; i >= 0; i--) {
        console.log(as[i], bs[i], i);

        if (as[i] === undefined) {
            sums.unshift(bs[i]);
        }
        if (bs[i] === undefined) {
            sums.unshift(as[i]);
        }
        if (as[i] === '1' && bs[i] === '1') {
            console.log('i should equal 1', i);
            sums.unshift('10');
        }
        if (as[i] === '0' && bs[i] === '0') {
            sums.unshift('0');
        } else {
            sums.unshift('1');
        }
    }
    console.log(sums);
    for (let j = 0; j < sums.length; j++) {
        let cur = sums[j];
        let next = sums[j+1];
        if (cur === '0' && next === '10') {
            sums = sums.splice(j, 1);
        } else {
            result = result.concat(cur);
        }
    }
    return result;
};
// RAN out of time, first approach I hit an issue and came up with the one above
// below is a different persons approach
var addBinary = function (a, b) {
    let num1Arr = a.split("");
    let num2Arr = b.split("");
    let result = [];

    let temp = 0;

    let j = num2Arr.length - 1;
    let i = num1Arr.length - 1;
    console.log(num1Arr, num2Arr);
    while (i >= 0 || j >= 0 || temp > 0) {

        let value = (i >= 0 ? Number(num1Arr[i]) : 0) + (j >= 0 ? Number(num2Arr[j]) : 0) + temp;

        console.log(i, j, "vals from arrs 1, then 2.", num1Arr[i], num2Arr[j], value, 'temp,', temp);

        if (value === 2) {
            result.push(0);
            temp = 1;
        }
        else if (value === 1) {
            result.push(1);
            temp = 0;
        }
        else if (value === 0) {
            result.push(0);
            temp = 0;
        }
        else if (value === 3) {
            result.push(1);
            temp = 1;
        }
        i--;
        j--;
        console.log("res", result);
    }

    if (temp > 0) {
        result.push(temp);
    }
    console.log(result);
    return result.reverse().join('');
};
// Example 1:
// Input: a = "11", b = "1"
// Output: "100"
// console.log(addBinary('11', '1'), '100');
// Example 2:
// Input: a = "1010", b = "1011"
// Output: "10101"
// console.log(addBinary('1010', '1011'), '10101');
//  1010
// 11100
//100110
console.log(addBinary('1010', '11100'), '100110');

// split IP arrs on empty space
// create carry var = 0
// get lengths of both IP arrs
// loop backwards, kill if i OR j >=0  OR (also carry >=0)... not sure why
    // value = Number at arrs being greater than zero == the number at array.  + carry val

File: nishuProbs/easy/add_binary.py
# Given two binary strings a and b, return their sum as a binary string.

# Example 1:
# Input: a = "11", b = "1"
# Output: "100"

# Example 2:
# Input: a = "1010", b = "1011"
# Output: "10101"


# Constraints:

# 1 <= a.length, b.length <= 104
# a and b consist only of '0' or '1' characters.
# Each string does not contain leading zeros except for the zero itself.

# 1 + 1 = 10
# hold temp '10'
# 1 + 1 + temp = '11
# pad the shorter with zeros
# loop from the back of each num

class Solution:
    def addBinary(self, a: str, b: str) -> str:
        sum = ''
        max_len = max(len(a),len(b))
        # a = int(a)
        # b = int(b)
        a = a.zfill(max_len)
        b = b.zfill(max_len)

        temp = '0'
        print(a, b)
        for i in range(max_len - 1, -1, -1):
            print(i, 'sum', sum,'a', a[i], 'b', b[i], 'temp', temp,)
            cur_a = a[i]
            cur_b = b[i]
            cur_single = '0'
            # 1 + 1 + temp=1
            if cur_a == '1' and cur_b == '1' and temp == '1':
                temp = '1'
                cur_single = '1'
            # 1 + 0 and 0 + 1, temp=1
            elif cur_a == '1' and cur_b == '0' and temp == '1' or cur_a == '0' and cur_b == '1' and temp == '1':
                temp = '1'
                cur_single = '0'
            # 0+0 + temp=1
            elif cur_a == '0' and cur_b == '0' and temp == '1':
                temp = '0'
                cur_single = '1'

         # 1 + 1 + temp=0
            elif cur_a == '1' and cur_b == '1' and temp == '0':
                temp = '1'  # Carry is set to '1'
                cur_single = '0'  # Result of the addition is '0'

            # 1 + 0 and 0 + 1, temp=0
            elif cur_a == '1' and cur_b == '0' and temp == '0' or cur_a == '0' and cur_b == '1' and temp == '0':
                temp = '0'
                cur_single = '1'
            # 0+0 + temp=0
            elif cur_a == '0' and cur_b == '0' and temp == '0':
                temp = 0
                cur_single = '0'

            # adding to sum and end edge case of adding a one to front
            if i == 0 and temp == '1':
                sum = temp + cur_single + sum
            else:
                sum = cur_single + sum

        return sum

t = Solution()
# print(t.addBinary('11','1'), '100')
# print(t.addBinary('1010','1011'), '10101')
a92 = '100'
b92 = "110010"
# print(t.addBinary(a92, b92), "110110")
class Sol:
    def addBinary(self, a: str, b: str) -> str:
        # Align the numbers
        max_len = max(len(a), len(b))
        a = a.zfill(max_len)
        b = b.zfill(max_len)

        # Initialize variables
        sum = ''
        carry = 0

        # Iterate through the digits
        for i in range(max_len - 1, -1, -1):
            print('carry', carry)
            digit_sum = int(a[i]) + int(b[i]) + carry
            sum_digit = digit_sum % 2  # Determine the sum
            carry = digit_sum // 2  # Update the carry
            print('digit_sum', digit_sum, 'new carry', carry, 'sum_digit', sum_digit)

            sum = str(sum_digit) + sum  # Append the result
            print('new sum,', sum)
        # Handle final carry
        if carry == 1:
            sum = '1' + sum

        return sum

gpt = Sol()
print(gpt.addBinary(a92, b92), "110110")
